pageserver/src/tenant/storage_layer/merge_iterator.rs

```diff
@@ -0,0 +1,301 @@
+use std::{
+    cmp::Ordering,
+    collections::{binary_heap, BinaryHeap},
+};
+
+use bytes::Bytes;
+use pageserver_api::key::Key;
+use utils::lsn::Lsn;
+
+use crate::{context::RequestContext, repository::Value};
+
+use super::{
+    delta_layer::{DeltaLayerInner, DeltaLayerIterator},
+    image_layer::{ImageLayerInner, ImageLayerIterator},
+};
+
+#[derive(Clone, Copy)]
+enum LayerRef<'a> {
+    Image(&'a ImageLayerInner),
+    Delta(&'a DeltaLayerInner),
+}
+
+impl<'a> LayerRef<'a> {
+    fn iter(self, ctx: &'a RequestContext) -> LayerIterRef<'a> {
+        match self {
+            Self::Image(x) => LayerIterRef::Image(x.iter(ctx)),
+            Self::Delta(x) => LayerIterRef::Delta(x.iter(ctx)),
+        }
+    }
+}
+
+enum LayerIterRef<'a> {
+    Image(ImageLayerIterator<'a>),
+    Delta(DeltaLayerIterator<'a>),
+}
+
+impl LayerIterRef<'_> {
+    async fn next(&mut self) -> anyhow::Result<Option<(Key, Lsn, Value)>> {
+        match self {
+            Self::Delta(x) => x.next().await,
+            Self::Image(x) => x.next().await,
+        }
+    }
+}
+
+struct IteratorWrapper<'a> {
+    /// The potential next key of the iterator. If the layer is not loaded yet, it will be the start key encoded in the layer file.
+    /// Otherwise, it is the next key of the real iterator.
+    peek_next_value: Option<(Key, Lsn, Value)>,
+    layer: LayerRef<'a>,
+    ctx: &'a RequestContext,
+    iter: Option<LayerIterRef<'a>>,
+}
+
+impl<'a> std::cmp::PartialEq for IteratorWrapper<'a> {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl<'a> std::cmp::Eq for IteratorWrapper<'a> {}
+
+impl<'a> std::cmp::PartialOrd for IteratorWrapper<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<'a> std::cmp::Ord for IteratorWrapper<'a> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        use std::cmp::Ordering;
+        let a = self.peek_next_key_lsn();
+        let b = other.peek_next_key_lsn();
+        match (a, b) {
+            (Some((k1, l1)), Some((k2, l2))) => {
+                let loaded_1 = if self.is_loaded() { 1 } else { 0 };
+                let loaded_2 = if other.is_loaded() { 1 } else { 0 };
+                // when key_lsn are the same, the unloaded iter will always appear before the loaded one.
+                (k1, l1, loaded_1).cmp(&(k2, l2, loaded_2))
+            }
+            (Some(_), None) => Ordering::Less,
+            (None, Some(_)) => Ordering::Greater,
+            (None, None) => Ordering::Equal,
+        }
+        .reverse()
+    }
+}
+
+impl<'a> IteratorWrapper<'a> {
+    pub fn create_from_image_layer(
+        image_layer: &'a ImageLayerInner,
+        ctx: &'a RequestContext,
+    ) -> Self {
+        Self {
+            peek_next_value: Some((
+                image_layer.key_range().start,
+                image_layer.lsn(),
+                Value::Image(Bytes::new()),
```

Comment:
I think the entire `IteratorWrapper` is probably best modelled as an enum, to model the two states of not-loaded and loaded.

I.e., from

```rust
struct IteratorWrapper<'a> {
    /// The potential next key of the iterator. If the layer is not loaded yet, it will be the start key encoded in the layer file.
    /// Otherwise, it is the next key of the real iterator.
    peek_next_value: Option<(Key, Lsn, Value)>,
    layer: LayerRef<'a>,
    ctx: &'a RequestContext,
    iter: Option<LayerIterRef<'a>>,
}
```

to

```rust
enum IteratorWrapper<'a> {
    NotLoaded {
        ctx: &'a RequestContext,
        layer: LayerRef<'a> },
    Loaded { 
        ctx: &'a RequestContext,
        peeked: (Key, Lsn, Value),
        iter: LayerIterRef<'a>,
    }
}
```

And in a later step, abstract away the peeking, either by switching to `futures::Stream` so we get `futures::Peekable` out of the box, or by implementing our own `PeekableLayerIterRef<'a>`.